src/core/AutomationPattern.cpp

```diff
@@ -417,33 +613,31 @@ float *AutomationPattern::valuesAfter( const TimePos & _time ) const
 
 
 
-void AutomationPattern::flipY( int min, int max )
+void AutomationPattern::flipY(int min, int max)
 {
-	timeMap tempMap = m_timeMap;
-	timeMap::ConstIterator iterate = m_timeMap.lowerBound(0);
-	float tempValue = 0;
+	QMutexLocker m(&m_patternMutex);
 
-	int numPoints = 0;
+	timeMap::iterator it = m_timeMap.lowerBound(0);
 
-	for( int i = 0; ( iterate + i + 1 ) != m_timeMap.end() && ( iterate + i ) != m_timeMap.end() ; i++)
-	{
-		numPoints++;
-	}
+	if (it == m_timeMap.end()) { return; }
 
-	for( int i = 0; i <= numPoints; i++ )
-	{
+	float tempValue = 0;
+	float outValueOffset = 0;
 
-		if ( min < 0 )
+	do
+	{
+		if (min < 0)
 		{
-			tempValue = valueAt( ( iterate + i ).key() ) * -1;
-			putValue( TimePos( (iterate + i).key() ) , tempValue, false);
+			it.value() *= -1.0;
 		}
 		else
 		{
-			tempValue = max - valueAt( ( iterate + i ).key() );
-			putValue( TimePos( (iterate + i).key() ) , tempValue, false);
+			tempValue = max - valueAt(POS(it));
+			outValueOffset = OFFSET(it) * -1.0;
+			putValues(TimePos(POS(it)), tempValue, tempValue + outValueOffset, false, true);
 		}
-	}
+		++it;
+	} while (it != m_timeMap.end());
```

Comment:
It's possible to use a `for` loop.
```cpp
for (auto it = m_timeMap.lowerBound(0); it != m_timeMap.end(); ++it)
```
And why did you use `lowerBound(0)` instead of `begin()`? We don't allow points before time 0 normally(not sure if XML manipulation allows it).